share/availability/discovery/discovery.go

```diff
@@ -96,35 +117,74 @@ func (d *Discovery) WithOnPeersUpdate(f OnUpdatedPeers) {
 
 // handlePeersFound receives peers and tries to establish a connection with them.
 // Peer will be added to PeerCache if connection succeeds.
-func (d *Discovery) handlePeerFound(ctx context.Context, topic string, peer peer.AddrInfo) {
-	if peer.ID == d.host.ID() || len(peer.Addrs) == 0 || d.set.Contains(peer.ID) {
-		return
-	}
-	err := d.set.TryAdd(peer.ID)
-	if err != nil {
-		log.Debug(err)
-		return
+func (d *Discovery) handlePeerFound(ctx context.Context, cancelFind context.CancelFunc, peer peer.AddrInfo) bool {
+	log := log.With("peer", peer.ID)
+	switch {
+	case peer.ID == d.host.ID():
+		log.Debug("skip handle: self discovery")
+		return false
+	case len(peer.Addrs) == 0:
+		log.Debug("skip handle: empty address list")
+		return false
+	case d.set.Size() >= d.set.Limit():
+		log.Debug("skip handle: enough peers found")
+		return false
+	case d.connector.HasBackoff(peer.ID):
+		log.Debug("skip handle: backoff")
+		return false
 	}
 
-	err = d.connector.Connect(ctx, peer)
-	if err != nil {
-		log.Debug(err)
-		d.set.Remove(peer.ID)
-		return
+	switch d.host.Network().Connectedness(peer.ID) {
+	case network.Connected:
+		if !d.set.Add(peer.ID) {
+			log.Debug("skip handle: peer is already in discovery set")
+			return false
+		}
+		// notify our subscribers
+		d.onUpdatedPeers(peer.ID, true)
+		log.Debug("added peer to set")
+		// check if we should cancel discovery
+		if d.set.Size() >= d.set.Limit() {
+			log.Infow("soft peer limit reached", "count", d.set.Size())
+			cancelFind()
+		}
+		// we still have to backoff the connected peer
+		d.connector.Backoff(peer.ID)
+	case network.NotConnected:
+		d.connectingLk.Lock()
+		if _, ok := d.connecting[peer.ID]; ok {
+			d.connectingLk.Unlock()
+			log.Debug("skip handle: connecting to the peer in another routine")
+			return false
+		}
+		d.connecting[peer.ID] = cancelFind
+		d.connectingLk.Unlock()
```

Comment:
@Wondertan  Looks like there is a shadowed read issue that could happen here. Here is example of events sequence that would lead to bug:

1. check connectivity state in `handlePeer` -> NotConnected 
2. connect event in sub handler -> skipped, since peer not in `connecting` pool
3. added to connecting pool in highlighted code.
4.??? (no connecting event, since it is already skipped)

To fix it we need to add to connecting pool before connectivity check. Do you mind if I make a quick fix?

Reply:
I don't mind, please do